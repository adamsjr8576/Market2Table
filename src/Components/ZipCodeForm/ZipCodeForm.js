import React, { Component } from 'react';
import './ZipCodeForm.scss';

class ZipCodeForm extends Component {
  constructor() {
    super()
    this.state = {
      zipCode: ''
    }
  }

  handleZipCodeChange = (e) => {
    if (e.target.value.length > 5) {
      e.target.value = e.target.value.slice(0, 5);
    }
    this.setState({ [e.target.name]: e.target.value })
  }

  render() {
    const opacity = 0;
    return (
      <main className="zip-code-container" style={{backgroundImage: `linear-gradient(rgba(255, 255, 255, ${opacity}), rgba(255, 255, 255, ${opacity}))` }}>
        <form className="zip-code-form">
          <label className="zip-code-input-label" htmlFor="zipCode">
            Enter zip code to find farmers markets near you:
          </label>
          <input onChange={(e) => this.handleZipCodeChange(e)} value={this.state.zipCode} className="zip-code-input" type="number" placeholder="Zip Code..." id="zipCode" name="zipCode"/>
          <button className="zip-code-button" type="button">Find!</button>
        </form>
      </main>
    )
  }
}

export default ZipCodeForm;
