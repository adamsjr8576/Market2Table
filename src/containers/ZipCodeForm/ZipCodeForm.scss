.zip-code-container {
  width: 100%;
  height: 100vh;
  position: fixed;
  z-index: 100;
  display: flex;
  align-items: center;
  justify-content: center;
}

.zip-code-form {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-around;
  height: 225px;
  width: 600px;
  color: white;
  border: 3px solid white;
  border-radius: 30px;
  background-image: linear-gradient(rgba(0, 0, 0, .7), rgba(0, 0, 0, .7));
}

.zip-code-input-label {
  font-size: 1.5em;
}

.zip-code-input {
  font-size: 1em;
  padding: 10px;
  border-radius: 20px;
  border: 0;
  outline: 0;
  appearance: none;
}

.zip-code-input:focus {
  transform: scale(1.05);
}

input::-webkit-outer-spin-button,
input::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

.zip-code-button {
  font-size: 1.3em;
  padding: 5px;
  border-radius: 10px;
  width: 20%;
  cursor: pointer;
  outline: 0;
}

.zip-code-button:hover,
.zip-code-button:focus {
  transform: scale(1.1);
}
