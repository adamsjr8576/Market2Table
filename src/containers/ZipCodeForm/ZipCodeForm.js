import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';
import './ZipCodeForm.scss';
import { farmersMarkets, marketsInfo } from '../../mockData.js';
import { addMarkets, addZipCode, addFavorites } from '../../actions/index.js';

export class ZipCodeForm extends Component {
  constructor() {
    super()
    this.state = {
      zipCode: '',
      hasError: false
    }
  }

  componentDidMount = () => {
    var marketsFromStorage = localStorage.getItem('favorites');
    var parsedMarkets = JSON.parse(marketsFromStorage);
    if (parsedMarkets !== null) {
      this.props.addFavorites(parsedMarkets);
    }
  }

  handleZipCodeChange = (e) => {
    if (e.target.value.length > 5) {
      e.target.value = e.target.value.slice(0, 5);
    }
    this.setState({ [e.target.name]: e.target.value })
  }

  handleZipCodeSubmit = favorites => {
    const { addMarkets, addZipCode } = this.props;
    if(this.state.zipCode.length === 5) {
      fetch(process.env.REACT_APP_BACKEND_URL + `/v1/data.svc/zipSearch?zip=${this.state.zipCode}`)
        .then(res => res.json())
        .then(markets => {
          let marketPromises = markets.results.map(market => {
            return fetch(process.env.REACT_APP_BACKEND_URL + `/v1/data.svc/mktDetail?id=${market.id}`)
              .then(res =>  res.json())
              .then(marketInfo => {
                let marketNameSplit = market.marketname.split(' ');
                if (marketNameSplit[0].includes('.')) {
                  marketNameSplit.shift();
                  marketNameSplit = marketNameSplit.join(' ');
                }
                const split1 = marketInfo.marketdetails.GoogleLink.split('=').pop();
                const split2 = split1.split('%');
                const lat = split2[0];
                const long = split2[2].slice(2);
                return {
                id: market.id,
                favorite: false,
                marketname: marketNameSplit,
                Address: marketInfo.marketdetails.Address,
                GoogleLink: marketInfo.marketdetails.GoogleLink,
                Products: marketInfo.marketdetails.Products,
                Schedule: marketInfo.marketdetails.Schedule,
                latitude: lat,
                longitude: long
                }
              })
              .catch(err => console.log(err))
          });
          return Promise.all(marketPromises);
        })
        .then(allMarketInfo => {
          const updatedFavorites = allMarketInfo.map(market => {
            if (favorites.length) {
              const objectCheck = favorites.find(favorite => favorite.id === market.id);
              if (objectCheck) {
                return objectCheck
              } else {
                return market
              }
            } else {
              return market
            }
          });
          addMarkets(updatedFavorites);
        })
        .catch(err => console.log(err))
      addZipCode(this.state.zipCode);
    }
    this.setState({ zipCode: ''});
  }

  render() {
    console.log(process.env.REACT_APP_API_KEY)
    let opacity = 0;
    if(this.props.path.includes('markets')) {
      opacity = .9;
    }
    if (this.props.zipCode.length > 0) {
      return (
        <Redirect to='/markets' />
      )
    }
    return (
      <main className="zip-code-container" style={{backgroundImage: `linear-gradient(rgba(255, 255, 255, ${opacity}), rgba(255, 255, 255, ${opacity}))` }}>
        <form className="zip-code-form">
          <label className="zip-code-input-label" htmlFor="zipCode">
            Enter zip code to find farmers markets near you:
          </label>
          <input onChange={(e) => this.handleZipCodeChange(e)} value={this.state.zipCode} className="zip-code-input" type="number" placeholder="Zip Code..." id="zipCode" name="zipCode"/>
          <button onClick={() => this.handleZipCodeSubmit(this.props.favorites)} className="zip-code-button" type="button">Find!</button>
        </form>
      </main>
    )
  }
}

export const mapDispatchToProps = dispatch => ({
  addMarkets: markets => dispatch( addMarkets(markets) ),
  addZipCode: zipCode => dispatch( addZipCode(zipCode) ),
  addFavorites: favorites => dispatch( addFavorites(favorites) )
});

export const mapStateToProps = state => ({
  zipCode: state.zipCode,
  favorites: state.favorites
});

export default connect(mapStateToProps, mapDispatchToProps)(ZipCodeForm);

ZipCodeForm.propTypes = {
  addMarkets: PropTypes.func,
  addZipCode: PropTypes.func,
  addFavorites: PropTypes.func,
  zipCode: PropTypes.string,
  favorites: PropTypes.array,
  path: PropTypes.string
}
