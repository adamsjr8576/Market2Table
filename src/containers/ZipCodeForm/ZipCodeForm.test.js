import React from 'react';
import { shallow } from 'enzyme';
import { ZipCodeForm, mapDispatchToProps, mapStateToProps } from './ZipCodeForm';
import { addMarkets, addZipCode, addFavorites } from '../../actions/index.js';

describe('ZipCodeForm', () => {
  describe('ZipCodeForm Component', () => {
    let mockAddMarkets;
    let mockAddZipCode;
    let mockAddFavorites;
    let mockZipCode;
    let mockFavorites;
    let mockPath;

    beforeEach(() => {
      mockAddMarkets = jest.fn();
      mockAddZipCode = jest.fn();
      mockAddFavorites = jest.fn();
      mockFavorites = [];
    });

    it('should match the snapshot with path of markets and the right zipcode', () => {
      mockZipCode = '80401';
      mockPath = '/markets';
      const wrapper = shallow(
        <ZipCodeForm
          addMarkets={mockAddMarkets}
          addZipCode={mockAddZipCode}
          addFavorites={mockAddFavorites}
          zipCode={mockZipCode}
          favorites={mockFavorites}
          path={mockPath}
        />);

      expect(wrapper).toMatchSnapshot();
    });

    it('should match the snapshot with path of markets and no zipcode', () => {
      mockZipCode = '';
      mockPath = '/markets';
      const wrapper = shallow(
        <ZipCodeForm
          addMarkets={mockAddMarkets}
          addZipCode={mockAddZipCode}
          addFavorites={mockAddFavorites}
          zipCode={mockZipCode}
          favorites={mockFavorites}
          path={mockPath}
        />);

      expect(wrapper).toMatchSnapshot();
    });

    it('should match the snapshot with favorites path and no zipcode', () => {
      mockZipCode = '';
      mockPath = '/favorites';
      const wrapper = shallow(
        <ZipCodeForm
          addMarkets={mockAddMarkets}
          addZipCode={mockAddZipCode}
          addFavorites={mockAddFavorites}
          zipCode={mockZipCode}
          favorites={mockFavorites}
          path={mockPath}
        />);

      expect(wrapper).toMatchSnapshot();
    });

    it('Should have a default state', () => {
      mockZipCode = '';
      mockPath = '/favorites';
      const wrapper = shallow(
        <ZipCodeForm
          addMarkets={mockAddMarkets}
          addZipCode={mockAddZipCode}
          addFavorites={mockAddFavorites}
          zipCode={mockZipCode}
          favorites={mockFavorites}
          path={mockPath}
        />);
      const mockDefaultState = {
        zipCode: '',
        hasError: false
      }

      expect(wrapper.state()).toEqual(mockDefaultState);
    });

    it('should change state when handleZipCodeChange is invoked', () => {
      mockZipCode = '';
      mockPath = '/favorites';
      const mockEvent = {
        target: {
          name: 'zipCode',
          value: '80'
        }
      }
      const wrapper = shallow(
        <ZipCodeForm
          addMarkets={mockAddMarkets}
          addZipCode={mockAddZipCode}
          addFavorites={mockAddFavorites}
          zipCode={mockZipCode}
          favorites={mockFavorites}
          path={mockPath}
        />);
      let mockState = {
        zipCode: '80',
        hasError: false
      }
      wrapper.instance().handleZipCodeChange(mockEvent);
      expect(wrapper.state()).toEqual(mockState)
    });

    it('should change state when handleZipCodeSubmit is invoked', () => {
      mockZipCode = '80401';
      mockPath = '/favorites';
      const wrapper = shallow(
        <ZipCodeForm
          addMarkets={mockAddMarkets}
          addZipCode={mockAddZipCode}
          addFavorites={mockAddFavorites}
          zipCode={mockZipCode}
          favorites={mockFavorites}
          path={mockPath}
        />);
      let mockState = {
        zipCode: '',
        hasError: false
      }
      wrapper.setState({ zipCode: '80401', hasError: false });
      wrapper.instance().handleZipCodeSubmit(mockFavorites);
      expect(wrapper.state()).toEqual(mockState)
    });

    it('should invoke addMarkets with updatedFavorites when handleZipCodeSubmit is invoked', () => {
      mockZipCode = '80401';
      mockPath = '/favorites';
      const wrapper = shallow(
        <ZipCodeForm
          addMarkets={mockAddMarkets}
          addZipCode={mockAddZipCode}
          addFavorites={mockAddFavorites}
          zipCode={mockZipCode}
          favorites={mockFavorites}
          path={mockPath}
        />);
      const mockMarkets = [
        {"id":"1002192", "marketname":"Ballston FRESHFARM Market", "GoogleLink":"http://maps.google.com/?q=38.881112%2C%20-77.112179%20(%22Ballston+FRESHFARM+Market%22)",
        "Address":"901 N Taylor St, Ballston, Virginia, 22203",
        "Schedule":"June - October Thursday 3:00 PM to 7:00 PM",
        "Products":"Baked goods; Cheese and/or dairy products; Eggs; Fresh fruit and vegetables; Fresh and/or dried herbs; Honey; Meat; Poultry", "favorite": false, "latitude": "38.881112", "longitude": "-77.112179"},
        {"id":"22008839", "marketname":"Ballston Farmers Market", "GoogleLink":"http://maps.google.com/?q=38.881112%2C%20-77.112179%20(%22Ballston+FRESHFARM+Market%22)",
        "Address":"901 N Taylor St, Ballston, Virginia, 22203",
        "Schedule":"June - October Thursday 3:00 PM to 7:00 PM",
        "Products":"Baked goods; Cheese and/or dairy products; Eggs; Fresh fruit and vegetables; Fresh and/or dried herbs; Honey; Meat; Poultry", "favorite": false, "latitude": "38.881112", "longitude": "-77.112179"},
        {"id":"1002291", "marketname":"Four Mile Run Farmers & Artisans Market", "GoogleLink":"http://maps.google.com/?q=38.881112%2C%20-77.112179%20(%22Ballston+FRESHFARM+Market%22)",
        "Address":"901 N Taylor St, Ballston, Virginia, 22203",
        "Schedule":"June - October Thursday 3:00 PM to 7:00 PM",
        "Products":"Baked goods; Cheese and/or dairy products; Eggs; Fresh fruit and vegetables; Fresh and/or dried herbs; Honey; Meat; Poultry", "favorite": false, "latitude": "38.881112", "longitude": "-77.112179"},
        {"id":"1002181", "marketname":"Dupont Circle FRESHFARM Market", "GoogleLink":"http://maps.google.com/?q=38.881112%2C%20-77.112179%20(%22Ballston+FRESHFARM+Market%22)",
        "Address":"901 N Taylor St, Ballston, Virginia, 22203",
        "Schedule":"June - October Thursday 3:00 PM to 7:00 PM",
        "Products":"Baked goods; Cheese and/or dairy products; Eggs; Fresh fruit and vegetables; Fresh and/or dried herbs; Honey; Meat; Poultry", "favorite": false, "latitude": "38.881112", "longitude": "-77.112179"}
      ]
      wrapper.setState({ zipCode: '80401' });
      wrapper.instance().handleZipCodeSubmit(mockFavorites);
      expect(mockAddMarkets).toHaveBeenCalledWith(mockMarkets);
      expect(mockAddZipCode).toHaveBeenCalledWith(wrapper.state('zipCode'));
    });

    it('Should invoke handleZipCodeChange on input change', () => {
      mockZipCode = '';
      mockPath = '/favorites';
      const wrapper = shallow(
        <ZipCodeForm
          addMarkets={mockAddMarkets}
          addZipCode={mockAddZipCode}
          addFavorites={mockAddFavorites}
          zipCode={mockZipCode}
          favorites={mockFavorites}
          path={mockPath}
        />);
        wrapper.instance().handleZipCodeChange = jest.fn();
        const mockEvent = {
          target: {
            name: 'zipCode',
            value: '80'
          }
        }
        wrapper.find('.zip-code-input').simulate('change', mockEvent);
        expect(wrapper.instance().handleZipCodeChange).toHaveBeenCalledWith(mockEvent)
    });

    it('Should invoke handleZipCodeSubmit on button click', () => {
      mockZipCode = '';
      mockPath = '/favorites';
      const wrapper = shallow(
        <ZipCodeForm
          addMarkets={mockAddMarkets}
          addZipCode={mockAddZipCode}
          addFavorites={mockAddFavorites}
          zipCode={mockZipCode}
          favorites={mockFavorites}
          path={mockPath}
        />);
        wrapper.instance().handleZipCodeSubmit = jest.fn();
        wrapper.find('.zip-code-button').simulate('click');
        expect(wrapper.instance().handleZipCodeSubmit).toHaveBeenCalledWith(mockFavorites)
    });
  });

  describe('mapStateToProps', () => {
    it('Should return an object with the zipCode and favorites', () => {
      const mockState = {
        farmersMarkets:
        [
          {
            id: "1002192",
            marketname: "Ballston FRESHFARM Market",
            favorite: false,
            GoogleLink: "http://maps.google.com/?q=38.881112%2C%20-77.112179%20(%22Ballston+FRESHFARM+Market%22)",
            Address: "901 N Taylor St, Ballston, Virginia, 22203",
            Schedule: "June - October Thursday 3:00 PM to 7:00 PM",
            Products: "Baked goods; Cheese and/or dairy products; Eggs; Fresh fruit and vegetables; Fresh and/or dried herbs; Honey; Meat; Poultry",
            latitude: "38.881112",
            longitude: "-77.112179"
          },
          {
            id: "22008839",
            marketname: "Ballston Farmers Market",
            favorite: false,
            GoogleLink: "http://maps.google.com/?q=38.881112%2C%20-77.112179%20(%22Ballston+FRESHFARM+Market%22)",
            Address: "901 N Taylor St, Ballston, Virginia, 22203",
            Schedule: "June - October Thursday 3:00 PM to 7:00 PM",
            Products: "Baked goods; Cheese and/or dairy products; Eggs; Fresh fruit and vegetables; Fresh and/or dried herbs; Honey; Meat; Poultry",
            latitude: "38.881112",
            longitude: "-77.112179"
          }],
        favorites: [],
        zipCode: '80401'
      }
      const expected = {
        favorites: [],
        zipCode: '80401'
      }
      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expected);
    });
  });

  describe('mapDispatchToProps', () => {
    it('Should call dispatch with addMarkets', () =>{
      const mockMarkets = [
        {"id":"1002192", "marketname":"Ballston FRESHFARM Market", "GoogleLink":"http://maps.google.com/?q=38.881112%2C%20-77.112179%20(%22Ballston+FRESHFARM+Market%22)",
        "Address":"901 N Taylor St, Ballston, Virginia, 22203",
        "Schedule":"June - October Thursday 3:00 PM to 7:00 PM",
        "Products":"Baked goods; Cheese and/or dairy products; Eggs; Fresh fruit and vegetables; Fresh and/or dried herbs; Honey; Meat; Poultry", "favorite": false, "latitude": "38.881112", "longitude": "-77.112179"},
        {"id":"22008839", "marketname":"Ballston Farmers Market", "GoogleLink":"http://maps.google.com/?q=38.881112%2C%20-77.112179%20(%22Ballston+FRESHFARM+Market%22)",
        "Address":"901 N Taylor St, Ballston, Virginia, 22203",
        "Schedule":"June - October Thursday 3:00 PM to 7:00 PM",
        "Products":"Baked goods; Cheese and/or dairy products; Eggs; Fresh fruit and vegetables; Fresh and/or dried herbs; Honey; Meat; Poultry", "favorite": false, "latitude": "38.881112", "longitude": "-77.112179"},
        {"id":"1002291", "marketname":"Four Mile Run Farmers & Artisans Market", "GoogleLink":"http://maps.google.com/?q=38.881112%2C%20-77.112179%20(%22Ballston+FRESHFARM+Market%22)",
        "Address":"901 N Taylor St, Ballston, Virginia, 22203",
        "Schedule":"June - October Thursday 3:00 PM to 7:00 PM",
        "Products":"Baked goods; Cheese and/or dairy products; Eggs; Fresh fruit and vegetables; Fresh and/or dried herbs; Honey; Meat; Poultry", "favorite": false, "latitude": "38.881112", "longitude": "-77.112179"},
        {"id":"1002181", "marketname":"Dupont Circle FRESHFARM Market", "GoogleLink":"http://maps.google.com/?q=38.881112%2C%20-77.112179%20(%22Ballston+FRESHFARM+Market%22)",
        "Address":"901 N Taylor St, Ballston, Virginia, 22203",
        "Schedule":"June - October Thursday 3:00 PM to 7:00 PM",
        "Products":"Baked goods; Cheese and/or dairy products; Eggs; Fresh fruit and vegetables; Fresh and/or dried herbs; Honey; Meat; Poultry", "favorite": false, "latitude": "38.881112", "longitude": "-77.112179"}
      ]
      const mockDispatch = jest.fn();
      const actionToDispatch = addMarkets(mockMarkets);
      const mappedProps = mapDispatchToProps(mockDispatch);

      mappedProps.addMarkets(mockMarkets);
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('Should call dispatch with addZipCode', () =>{
      const mockZipCode = '80401';
      const mockDispatch = jest.fn();
      const actionToDispatch = addZipCode(mockZipCode);
      const mappedProps = mapDispatchToProps(mockDispatch);

      mappedProps.addZipCode(mockZipCode);
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });

    it('Should call dispatch with addFavorites', () =>{
      const mockfavorites = [];
      const mockDispatch = jest.fn();
      const actionToDispatch = addFavorites(mockfavorites);
      const mappedProps = mapDispatchToProps(mockDispatch);

      mappedProps.addFavorites(mockfavorites);
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
    });
  });
});
