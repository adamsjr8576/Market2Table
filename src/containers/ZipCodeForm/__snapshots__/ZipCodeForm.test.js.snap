// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ZipCodeForm ZipCodeForm Component should match the snapshot with favorites path and no zipcode 1`] = `
<main
  className="zip-code-container"
  style={
    Object {
      "backgroundImage": "linear-gradient(rgba(255, 255, 255, 0), rgba(255, 255, 255, 0))",
    }
  }
>
  <form
    className="zip-code-form"
  >
    <label
      className="zip-code-input-label"
      htmlFor="zipCode"
    >
      Enter zip code to find farmers markets near you:
    </label>
    <input
      className="zip-code-input"
      id="zipCode"
      name="zipCode"
      onChange={[Function]}
      placeholder="Zip Code..."
      type="number"
      value=""
    />
    <button
      className="zip-code-button"
      onClick={[Function]}
      type="button"
    >
      Find!
    </button>
  </form>
</main>
`;

exports[`ZipCodeForm ZipCodeForm Component should match the snapshot with path of markets and no zipcode 1`] = `
<main
  className="zip-code-container"
  style={
    Object {
      "backgroundImage": "linear-gradient(rgba(255, 255, 255, 0.9), rgba(255, 255, 255, 0.9))",
    }
  }
>
  <form
    className="zip-code-form"
  >
    <label
      className="zip-code-input-label"
      htmlFor="zipCode"
    >
      Enter zip code to find farmers markets near you:
    </label>
    <input
      className="zip-code-input"
      id="zipCode"
      name="zipCode"
      onChange={[Function]}
      placeholder="Zip Code..."
      type="number"
      value=""
    />
    <button
      className="zip-code-button"
      onClick={[Function]}
      type="button"
    >
      Find!
    </button>
  </form>
</main>
`;

exports[`ZipCodeForm ZipCodeForm Component should match the snapshot with path of markets and the right zipcode 1`] = `
<Redirect
  to="/markets"
/>
`;
